Add Next.js boilerplate template to create-aptos-dapp

aeither opened the pull request:
Purpose
Next.js is the most Popular React Framework widely used by Web Developer.

This PR adds a Next.js boilerplate template to the existing create-aptod-dapp project, expanding the options for developers who prefer Next.js for their Aptod dApp development.

Changes Made
Added a new Next.js template in template folder
Implemented basic Next.js project structure with Aptos integration
Updated the CLI to include the Next.js option alongside existing Vite templates

Benefits
Provides more flexibility for developers by offering a popular React framework option
Aligns with the project's goal of simplifying Aptos dApp bootstrapping

Testing
This is a fork from Boilerplate Template so for testing is mostly the same except for some improvements and additions:

- Faucet UI: Users can clain Testnet Aptos Coin
- Improved Form with React-Hook-Form
- Small Fixes

0xmaayan reviewed (changes requested):
This is great, and thank you for this amazing contribution!

We've had internal discussions about providing server-side rendering (Next.js) template support, and we believe the best approach is to include an option in the wizard flow to choose between client-side (Vite) or server-side (Next.js) frameworks for each template.

It would be fantastic if you could modify this PR to support this feature and provide all the benefits you mentioned.

The developer user flow would be:

1. Run npx create-aptos-dapp.
2. Choose a template.
3. The next question would be to choose between a server-side (Next.js app) or client-side (Vue app) framework.
4. Internally, based on the framework choice, the tool generates the corresponding files.

We can start by implementing this option in the Boilerplate template, and once the PR is approved, we can extend it to each template in the tool.

aeither commented:
Hey @0xmaayan,

Update with the requested flow:
Now, after running `npx create-aptos-dapp` and choosing a template,
the next question is to choose between a server-side (Next.js app) or client-side (Vue app) framework.

Internally, based on the framework choice, the tool generates the corresponding files for the boilerplate implemented. For the other options, it doesn't show the server option.

aeither reviewed (commented):
Update with the requested flow:
Now, after running npx create-aptos-dapp and choosing a template,
the next question is to choose between a server-side (Next.js app) or client-side (Vue app) framework.

Internally, based on the framework choice, the tool generates the corresponding files for the boilerplate implemented. For the other options, it doesn't show the server option.

aeither commented:
Hey @0xmaayan ,

I went through all your comments. 
I rebased from main and removed everything extra I had added (surf, faucet, etc.). 
Now it should only include the necessary changes for Next.js. 

By Next.js convention, a ⁠src or ⁠app folder are the entry points for a Next.js app and must be located at the root level. If so, we need to move all the rest (package.json, tsconfig…) inside the frontend folder. At that point, the user will then need to ⁠cd into the frontend folder and run ⁠npm run dev.

Let me know if I overlooked anything

0xmaayan reviewed (commented):
Starts to look much better!

- Please revert those wild format changes =)
- Since the tool generates `env` vars for each template, we need to make sure it generates the required vars for a nextjs framework, i.e `NEXT_*`.  See my comment on adding a `framework` prop to each template

aeither commented on templates/nextjs-boilerplate-template/next.config.mjs:
```diff
@@ -0,0 +1,8 @@
+const nextConfig = {
+  output: "export", // Outputs a Single-Page Application (SPA).
+  distDir: "./dist", // Changes the build output directory to `./dist/`.
+  basePath: process.env.NEXT_PUBLIC_BASE_PATH, // Sets the base path to `/some-base-path`.
```
reporting nextjs docs "To deploy a Next.js application under a sub-path of a domain you can use the basePath config option."
still everything have to be moved to the frontend folder.

anyways i tried setting basePath to frontend and still got the error.
"
Error: > Couldn't find any `pages` or `app` directory. Please create one under the project root"

0xmaayan replied:
yea.. I guess it is a Nextjs hard requirement, thats fine! thanks for looking into it

aeither commented on src/workflow.ts:
```diff
@@ -1,78 +1,82 @@
-    // A boolean variable that keeps track on whether the user wants to change their initial choices
-    let confirmOptions = true;
+	try {
+		// A boolean variable that keeps track on whether the user wants to change their initial choices
```
i am using a different formatter.. anyways. it is reverted.

aeither commented on src/generateDapp.ts:
```diff
@@ -111,6 +111,9 @@ export async function generateDapp(selection: Selections) {
         await generateEnvFile();
         break;
+      case "nextjs-boilerplate-template":
+        await generateEnvFile();
```
now .env use nextjs env when using nextjs template

0xaptosj commented:
hi @aeither thanks a lot for the contribution, are you open to a chat? we are very impressed with your background and want to learn more about what you want to build on aptos, and happy to take any feedback. You can share your twitter if you want, then ill dm you my calendly

0xmaayan commented on .github/workflows/build-aptos-next-template.yml:
```diff
@@ -0,0 +1,31 @@
+name: Build Aptos Next Template Frontend
```
```suggestion
name: Build Nextjs Boilerplate Template Frontend
```

0xmaayan reviewed (commented):
I think we’re almost there! I’ve made a couple of small tweaks in this commit: https://github.com/aptos-labs/create-aptos-dapp/pull/203/commits/d5deef74e2092e62d31f94ca688ea8dcf8f307f8

Please include these changes in this PR.

aeither commented:
added your commit!

just one thing. I reverted NEXT_APP_NETWORK back to NEXT_PUBLIC_APP_NETWORK

because it is used in constants.ts

`export const NETWORK = process.env.NEXT_APP_NETWORK ?? "testnet";`

and exposed to client side

`<AptosWalletAdapterProvider
      autoConnect={true}
      dappConfig={{ network: NETWORK }}`
      
      anything env exposed to client side must be prefixed with NEXT_PUBLIC_ otherwise it is going to be undefined due to security reasons

0xmaayan commented on src/generateDapp.ts:
```diff
@@ -138,6 +138,9 @@ export async function generateDapp(selection: Selections) {
         await generateEnvFile();
         break;
+      case "nextjs-boilerplate-template":
+        await generateEnvFile();
```
I think you have lost this change after rebase https://github.com/aptos-labs/create-aptos-dapp/commit/d5deef74e2092e62d31f94ca688ea8dcf8f307f8#diff-0594d46782df78830c5efe0fa4abb36cbf22c918dc182ccc8f3cf75fdee908e7R85

aeither replied:
should be fixed

aeither commented:
> hi @aeither thanks a lot for the contribution, are you open to a chat? we are very impressed with your background and want to learn more about what you want to build on aptos, and happy to take any feedback. You can share your twitter if you want, then ill dm you my calendly

hey, my twtt @giovannifulin

0xaptosj commented:
> > hi @aeither thanks a lot for the contribution, are you open to a chat? we are very impressed with your background and want to learn more about what you want to build on aptos, and happy to take any feedback. You can share your twitter if you want, then ill dm you my calendly
> 
> hey, my twtt @giovannifulin

thanks i just followed you, i can't dm you on twitter, probably because your dm permission is no one can dm you (i think that's the default option), you can dm me on twitter, im `0xaptosj`

0xmaayan reviewed (commented):
Hi @aeither , please take a look at this commit and apply the changes to the PR
https://github.com/aptos-labs/create-aptos-dapp/pull/204/commits/7dab9bfeb320048cdbf775d5dc63ae0eb8f8e13f

0xmaayan commented:
@aeither build is breaking https://github.com/aptos-labs/create-aptos-dapp/actions/runs/10745190338/job/29884546340?pr=197
I'd suggest try building it locally and make sure it works

aeither commented:
@0xmaayan  fixed and deployed https://create-aptos-dapp.vercel.app/

0xmaayan commented:
@aeither we now have some format issues https://github.com/aptos-labs/create-aptos-dapp/actions/runs/10777656989/job/29887815914?pr=197

0xmaayan commented:
@aeither Merged! Thanks a lot for the amazing contribution and collaboration!
